File: LeetCode/1162. As Far from Land as Possible/solution.py
"""
37 / 37 test cases passed.
Runtime: 248 ms
Memory Usage: 15.7 MB
"""
class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        n = len(grid)
        que = collections.deque([(i, j) for i in range(n) for j in range(n) if grid[i][j]]);
        if not que or len(que) == n * n:
            return -1 
        ans = -1
        while que:
            ans += 1
            for _ in range(len(que)):
                x, y = que.popleft()
                for xx, yy in [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]:
                    if 0 <= xx < n and 0 <= yy < n and grid[xx][yy] == 0:
                        grid[xx][yy] = 1
                        que.append((xx, yy))
        return ans
            

File: LeetCode/1162. As Far from Land as Possible/solution.cpp
/**
 * 37 / 37 test cases passed.
 * Runtime: 52 ms
 * Memory Usage: 19.3 MB 
 */
class Solution {
public:
    int direct[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    using pii = pair<int, int>;
    int maxDistance(vector<vector<int>>& grid) {
        int n = grid.size();
        queue<pii> que;
        for (int i = 0; i < n; ++ i) {
            for (int j = 0; j < n; ++ j) {
                if (grid[i][j]) que.emplace(i, j);
            }
        }
        if (que.size() == 0 || que.size() == n * n) return -1;

        int x, y;
        while (!que.empty()) {
            auto temp = que.front(); que.pop();
            x = temp.first, y = temp.second;
            for (auto &d : direct) {
                int xx = x + d[0];
                int yy = y + d[1];
                if (xx < 0 || xx >= n || yy < 0 || yy >= n) continue;
                if (grid[xx][yy]) continue;
                grid[xx][yy] = grid[x][y] + 1;
                que.emplace(xx, yy);
            }
        }
        return grid[x][y] - 1;
    }
};
